tests/test_lgr.py

```diff
@@ -0,0 +1,159 @@
+        'My  s       Marko   poexa-l-i   avtobus-om  v   Peredelkino.',
+        '1PL COM     Marko   go-PST-PL   bus-INS     All Peredelkino.')
+    assert igt.gloss_abbrs['1PL'] == 'first person plural'
+
+    igt = assert_is_valid(
+        'My  s       Marko   poexa-l-i   avtobus-om  v   Peredelkino.',
+        '1PL COM     Marko   go-PST-PL   bus-INS     All Peredelkino.',
+        translation="'Marko and I went to Perdelkino by bus.' (COM=whatever)")
+    assert igt.gloss_abbrs['COM'] == 'whatever'
+
+
+def test_rule4():
+    assert_is_valid('çık-mak', 'come.out-INF')
+
+    igt = assert_is_valid('insul-arum', 'island-GEN.PL')
+    assert 'PL' in igt.gloss_abbrs
+
+    assert_is_valid('aux         chevaux', 'to.ART.PL   horse.PL')
+
+    assert_is_valid('unser-n     Väter-n', 'our-DAT.PL  father.PL-DAT.PL')
+
+    igt = assert_is_valid(
+        'n=an        apedani     mehuni      essandu.',
+        'CONN=him    that.DAT.SG time.DAT.SG eat.they.shall',
+        translation="'They shall celebrate him on that date.' (CONN = connective)")
+    assert igt.gloss_abbrs['CONN'] == 'connective'
+
+    assert_is_valid('nanggayan   guny-bi-yarluga?', 'who         2DU.A.3SG.P-FUT-poke')
+
+    igt = assert_is_valid('çık-mak', 'come_out-INF')
+    assert igt.glossed_morphemes[0][0][1] == 'come_out'
+
+    igt = assert_is_valid('insul-arum', 'island-GEN;PL')
+    assert 'PL' in igt.gloss_abbrs
+
+    igt = assert_is_valid('aux         chevaux', 'to;ART;PL   horse;PL')
+    assert 'ART' in igt.gloss_abbrs
+
+    igt = assert_is_valid(
+        'n=an        apedani     mehuni      essandu.',
+        'CONN=him    that:DAT;SG time:DAT;SG eat:they:shall')
+    assert 'DAT' in igt.gloss_abbrs
+
+    igt = assert_is_valid('unser-n     Väter-n', r'our-DAT.PL father\PL-DAT')
+    assert 'PL' in igt.gloss_abbrs
+
+    igt = assert_is_valid('bhris-is', r'PST\break-2SG')
+    assert 'PST' in igt.gloss_abbrs
+
+    igt = assert_is_valid('mú-kòrà', r'SBJV\1PL-work')
+    assert 'SBJV' in igt.gloss_abbrs
+
+    igt = assert_is_valid('nanggayan   guny-bi-yarluga?', 'who         2DU>3SG-FUT-poke')
+    assert '2DU' in igt.gloss_abbrs
+
+
+def test_rule5():
+    igt = assert_is_valid('and-iamo', 'go-PRS.1.PL')
+    assert '1' not in igt.gloss_abbrs
+
+    # Rule 5A is hard to support.
+    """
+    Rule 5A. (Optional)
+
+    Number and gender markers are very frequent in some languages, especially
+    when combined with person. Several authors therefore use non-capitalized
+    shortened abbreviations without a period. If this option is adopted, then the
+    second gloss is used in (21).
+
+    (21) Belhare
+    ne-e    a-khim-chi      n-yuNNa
+    DEM-LOC that.DAT.SG     3NSG-be.NPST
+    DEM-LOC 1sPOSS-house-PL 3ns-be.NPST
+    'Here are my houses.''
+    """
+
+
+def test_rule6():
+    igt1 = assert_is_valid('puer', 'boy[NOM.SG]')
+    igt2 = assert_is_valid('puer-Ø', 'boy-NOM.SG')
+    assert igt2.primary_text == 'puer'
+    assert igt1 == igt2
```

Comment:
@fmatter I have no idea how common these non-overt elements are. But what the code does now, is normalize to the format with `Ø` in object language. Then, glosses can be compared for equality - which seems nice.

Reply:
How secure can we be that linguists use the empty set symbol instead of the Ö-Symbol? This is a perfect confusable, and although they should know, linguists typically don't know their Unicode well enough here. So I'd try and have a check for the symbol, and correct it internally (as we do with `:` as a length marker in [CLTS](https://clts.clld.org)), or through some explicit error, or allow for both cases.

Reply:
Ah, good catch! Will adjust the symbol! I guess, pretending to be able to detect gloss equivalence beyond simple string equality, would again be too ambitious. So I'll skip this normalization.

Reply:
How awful: I just saw that the null-morpheme is IN the original text, then it does not make sense to normalize, as we cannot tell what we'll get in the original text (although it is unlikely to find an Ø that represents a sound in this form). But then one might want to mention this somewhere, that Ø would not be allowed here.